### ISSUE TITLE
openai tools don't work with streaming=True

### ISSUE THREAD

**Issue by Dador** [langchain-ai/langchain#13442]
### System Info

langchain: latest (0.0.336)

### Who can help?

@hwchase17 (from git blame and from #13110)

### Information

- [ ] The official example notebooks/scripts
- [X] My own modified scripts

### Related Components

- [X] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [ ] Agents / Agent Executors
- [ ] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

Code to reproduce (based on [code from docs](https://python.langchain.com/docs/modules/agents/agent_types/openai_tools))
```python
from langchain.agents import AgentExecutor
from langchain.agents.format_scratchpad.openai_tools import (
    format_to_openai_tool_messages,
)
from langchain.agents.output_parsers.openai_tools import OpenAIToolsAgentOutputParser
from langchain.chat_models import ChatOpenAI
from langchain.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain.tools import BearlyInterpreterTool, DuckDuckGoSearchRun
from langchain.tools.render import format_tool_to_openai_tool


lc_tools = [DuckDuckGoSearchRun()]
oai_tools = [format_tool_to_openai_tool(tool) for tool in lc_tools]

prompt = ChatPromptTemplate.from_messages(
    [
        ("system", "You are a helpful assistant"),
        ("user", "{input}"),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]
)
llm = ChatOpenAI(temperature=0, model="gpt-3.5-turbo-1106",
                 streaming=True)

agent = (
    {
        "input": lambda x: x["input"],
        "agent_scratchpad": lambda x: format_to_openai_tool_messages(
            x["intermediate_steps"]
        ),
    }
    | prompt
    | llm.bind(tools=oai_tools)
    | OpenAIToolsAgentOutputParser()
)

agent_executor = AgentExecutor(agent=agent, tools=lc_tools, verbose=True)

agent_executor.invoke(
    {"input": "What's the average of the temperatures in LA, NYC, and SF today?"}
)
```

Logs:
```
> Entering new AgentExecutor chain...
Traceback (most recent call last):
  File "./test-functions.py", line 42, in <module>
    agent_executor.invoke(
  File "./venv/lib/python3.11/site-packages/langchain/chains/base.py", line 87, in invoke
    return self(
           ^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/chains/base.py", line 310, in __call__
    raise e
  File "./venv/lib/python3.11/site-packages/langchain/chains/base.py", line 304, in __call__
    self._call(inputs, run_manager=run_manager)
  File "./venv/lib/python3.11/site-packages/langchain/agents/agent.py", line 1245, in _call
    next_step_output = self._take_next_step(
                       ^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/agents/agent.py", line 1032, in _take_next_step
    output = self.agent.plan(
             ^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/agents/agent.py", line 461, in plan
    output = self.runnable.invoke(inputs, config={"callbacks": callbacks})
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/schema/runnable/base.py", line 1427, in invoke
    input = step.invoke(
            ^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/schema/runnable/base.py", line 2787, in invoke
    return self.bound.invoke(
           ^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 142, in invoke
    self.generate_prompt(
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 459, in generate_prompt
    return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 349, in generate
    raise e
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 339, in generate
    self._generate_with_cache(
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 492, in _generate_with_cache
    return self._generate(
           ^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/openai.py", line 422, in _generate
    return _generate_from_stream(stream_iter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/chat_models/base.py", line 61, in _generate_from_stream
    generation += chunk
  File "./venv/lib/python3.11/site-packages/langchain/schema/output.py", line 94, in __add__
    message=self.message + other.message,
            ~~~~~~~~~~~~~^~~~~~~~~~~~~~~
  File "./venv/lib/python3.11/site-packages/langchain/schema/messages.py", line 225, in __add__
    additional_kwargs=self._merge_kwargs_dict(
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "./venv/lib/python3.11/site-packages/langchain/schema/messages.py", line 138, in _merge_kwargs_dict
    raise ValueError(
ValueError: Additional kwargs key tool_calls already exists in this message.
```

`left` and `right` from inside`_merge_kwargs_dict`:
```python
left = {'tool_calls': [{'index': 0, 'id': 'call_xhpbRSsUKkzsvtFgnkTXEFtAtHc', 'function': {'arguments': '', 'name': 'duckduckgo_search'}, 'type': 'function'}]}
right = {'tool_calls': [{'index': 0, 'id': None, 'function': {'arguments': '{"qu', 'name': None}, 'type': None}]}
```

### Expected behavior

No errors and same result as without `streaming=True`.

**Comment by JonatanMedinilla** [comment]
Hi guys, any news on this issue? I faced this issue today, and as the bot mentioned, it is an issue that only happens when we enable streaming in the OpenAI model. From what I could see, langchain is merging the message chunks, but when it tries to merge the additional_kwargs doesn't know how to handle the scenario when the key is **tool_calls**.

If you look at the sample log below, **tool_calls** is a **list** and that scenario is not being handled in the **BaseMessageChunk** class. 

`CHUNK in Stream: message=AIMessageChunk(content='', additional_kwargs={'tool_calls': [<OpenAIObject id=call_WUK4d3JwROWKKTfJfEEG6hpJ at 0x7fa71193f240> JSON: {
  "index": 0,
  "id": "call_WUK4d3JwROWKKTfJfEEG6hpJ",
  "type": "function",
  "function": {
    "name": "CreateLiveEvent",
    "arguments": ""
  }
}]})`

Next chunks will be like the next sample, and I think that the only way to know that this chunk relates to the previous one is by its index. 
`CHUNK in Stream: message=AIMessageChunk(content='', additional_kwargs={'tool_calls': [<OpenAIObject at 0x7fa71193c040> JSON: {
  "function": {
    "arguments": "{\n"
  },
  "index": 0
}]})`

I will try to work on a workaround by modifying the **BaseMessageChunk**, but I'm not sure if that's the correct approach. As far as I know/If I'm not wrong, the abstractions we have in langchain_core should not be coupled to a specific model, and it seems that this issue only happens with Open AI.  

Thoughts?

**Comment by JonatanMedinilla** [comment]
FYI, just modified the _merge_kwargs_dict method in the BaseMessageChunk class like this, to check if the value is a list and the key tool_calls. I know that it is not the best approach, but it helped me to workaround the issue in this scenario (It is probably that I'm missing edge cases due to my lack of context). Now, I'm able to merge the tool_calls, but the AsyncIteratorCallbackHandler is not returning any tokens.

```
def _merge_kwargs_dict(
        self, left: Dict[str, Any], right: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Merge additional_kwargs from another BaseMessageChunk into this one,
        handling specific scenarios where a key exists in both dictionaries
        but has a value of None in 'left'. In such cases, the method uses the
        value from 'right' for that key in the merged dictionary.
        Example:
        If left = {"function_call": {"arguments": None}} and
        right = {"function_call": {"arguments": "{\n"}}
        then, after merging, for the key "function_call",
        the value from 'right' is used,
        resulting in merged = {"function_call": {"arguments": "{\n"}}.
        """
        merged = left.copy()
        for k, v in right.items():
            if k not in merged:
                merged[k] = v
            elif merged[k] is None and v:
                merged[k] = v
            elif type(merged[k]) != type(v):
                
                raise ValueError(
                    f'additional_kwargs["{k}"] already exists in this message,'
                    " but with a different type."
                )
            elif isinstance(merged[k], str):
                merged[k] += v
            elif isinstance(merged[k], dict):
                merged[k] = self._merge_kwargs_dict(merged[k], v)
            elif isinstance(merged[k], list) and k == "tool_calls":
                merged_dict = {}
                for item in merged[k] + v:
                    index = item['index']
                    if index in merged_dict:
                        merged_dict[index]['function']['arguments'] += v[index]['function']['arguments']
                    else:
                        merged_dict[index] = item
                merged[k] = list(merged_dict.values())
            else:
                raise ValueError(
                    f"Additional kwargs key {k} already exists in this message."
                )
        return merged
```

**Comment by JonatanMedinilla** [comment]
I think this issue can be closed. @nfcampos created a PR with the fix and it was merged today.

Thanks Nuno! You rock! 👏

**Comment by Dador** [comment]
Fixed in #14613
